massa-execution/src/interface_impl.rs

```diff
@@ -29,15 +38,73 @@ impl Interface for InterfaceImpl {
     fn get_module(&self, address: &String) -> Result<Vec<u8>> {
         let bytecode = {
             let address = Address::from_str(address)?;
-            let context = self
-                .context
-                .lock()
-                .expect("Failed to acquire lock on context.");
+            let context = context_guard!(self);
             context.ledger_step.get_module(&address)
         };
         match bytecode {
             Some(bytecode) => Ok(bytecode),
             _ => bail!("Error bytecode not found"),
         }
     }
+
+    /// Requires a new address that contains the sent bytecode.
+    ///
+    /// Generate a new address with a concatenation of the firsts N bytes of the blockId and the firsts N bytes of the block creator address.
+    /// Where N = ADDRESS_SIZE_BYTES / 2
+    ///
+    /// Insert in the ledger the given bytecode in the generated address
+    fn create_module(&self, module: &Bytecode) -> Result<assembly_simulator::Address> {
+        let (block_id, index) = {
+            let context = context_guard!(self);
+            (context.opt_block_id, context.call_stack.len())
+        };
+        let (block_id, index) = match (block_id, u8::try_from(index)) {
+            (Some(block_id), Ok(index)) => (block_id, index),
+            _ => bail!("Failed to read current context"),
+        };
+        let mut data = block_id.to_bytes().to_vec();
+        data.push(index);
+        let address = Address(massa_hash::hash::Hash::from(&data));
+        let mut context = context_guard!(self);
+        context.ledger_step.set_module(address, module.clone());
+        Ok(address.to_string())
+    }
+
+    /// Requires the data at the address
+    fn get_data_for(&self, address: &assembly_simulator::Address, key: &str) -> Result<Bytecode> {
+        let context = context_guard!(self);
+        let addr = &Address::from_str(address)?;
+        let key = massa_hash::hash::Hash::from_bs58_check(key)?;
+        return match context.ledger_step.get_data_entry(addr, &key) {
+            Some(bytecode) => Ok(bytecode),
+            _ => bail!("Data entry not found"),
+        };
+    }
+
+    /// Requires to replace the data in the current address
+    ///
+    /// Note:
+    /// The execution lib will allways use the current context address for the update
+    fn set_data_for(
+        &self,
+        address: &assembly_simulator::Address,
+        key: &str,
+        value: &Bytecode,
+    ) -> Result<()> {
+        let mut context = context_guard!(self);
+        let addr = Address::from_str(address)?;
+        let key = massa_hash::hash::Hash::from_bs58_check(key)?;
+        if context.call_stack.contains(&addr) {
```

Comment:
access should be granted only if `addr == context.call_stack.back()` (you don't want something you call -possibly very indirectly- to write-access your own data) or if `context.created_addresses.contains(&addr)` (you can write-access SCs that you created during the same execution, useful for initialization)

Reply:
modified